test/ueberauth_adfs/adfs_test.exs

```diff
@@ -0,0 +1,352 @@
+defmodule Ueberauth.Strategy.ADFSTest do
+  use ExUnit.Case
+  use Ueberauth.Strategy
+
+  # Load ADFS Test Handler
+  Code.load_file("test/ueberauth_adfs/adfs_test_handler.exs")
+
+  import Mock
+
+  alias Ueberauth.Strategy.ADFS
+
+  @env_values adfs_url: "https://example.com",
+              adfs_metadata_url: "https://example.com/metadata.xml",
+              client_id: "example_client",
+              resource_identifier: "example_resource"
+
+
+  @env_handler_values adfs_url: "https://example.com",
+                      adfs_metadata_url: "https://example.com/metadata.xml",
+                      adfs_handler: Ueberauth.Strategy.ADFSTestHandler,
+                      client_id: "example_client",
+                      resource_identifier: "example_resource"
+
+  @mock_metadata {:ok,
+                  %HTTPoison.Response{
+                    body:
+                      "<EntityDescriptor><ds:Signature><KeyInfo>" <>
+                        "<X509Data><X509Certificate>1234</X509Certificate></X509Data>" <>
+                        "</KeyInfo></ds:Signature></EntityDescriptor>"
+                  }}
+
+  @user_claim %Joken.Token{
+    claims: %{
+      "email" => "user@test.com",
+      "given_name" => "John",
+      "family_name" => "Doe",
+      "winaccountname" => "john1"
+    }
+  }
+
+  describe "ADFS Strategy" do
+    setup_with_mocks [
```

Comment:
These mocks can get a bit messy but it's really tricky to test something like this otherwise.

Reply:
yeah they do look a little messy but it does seem like the best way to test something like this